src/nccl_ofi_rdma.cpp

```diff
@@ -5918,34 +5688,39 @@ static int send(nccl_net_ofi_send_comm_t *send_comm, void *data, size_t size, in
 		eager = true;
 	}
 
+	msg_seq_num = s_comm->next_msg_seq_num;
+	slot = msg_seq_num % NCCL_OFI_MAX_REQUESTS;
+
+	if (!eager) {
+		if (s_comm->ctrl_fifo[slot].id != ((uint64_t)(msg_seq_num + 1) & MSG_SEQ_NUM_MASK)) {
+			*base_req = NULL;
+			ret = 0;
+			goto error;
+		} else {
+			have_ctrl = true;
+			nccl_net_ofi_mutex_lock(&s_comm->ctrl_recv_lock);
+			s_comm->n_ctrl_received += 1;
```

Comment:
this makes me sad.  We're taking a lock to increment a single integer it's unclear we actually need to count this way.

On the other hand, you need a memory synchronization point here so that the compiler / processor doesn't start reading the data in the control message before the check for the id.